2021-10-08-reproducible-builds-for-debian-a-big-step-forward.md

```diff
@@ -1,42 +1,42 @@
+That's why we decided to create our own [snapshot](https://github.com/fepitre/debian-snapshot) service. Easy to say, but not to do. First, the original snapshot service from Debian is roughly 90TB of repository data. Second, we cannot download files easily because only HTTP(S) is available, and downloading multiple files means we are impeded by availability issues.
+In order to work around the huge volume of data, we decided to get repositories from 2017 to today (which corresponds approximately to when Debian "Buster" was released) and only related architectures `amd64`, `source`, and `all`. (`all` indicates no specific architecture in the Debian world.) For the download part itself, we needed to parse the metadata of each Debian repository in order to get the list of files to download for every timestamp for which a snapshot had been made. Then, we developed `resume` and `retry` download functions, which unfortunately are brute force download functions. For storing the data, a simple approach has been employed: storing files as SHA-256 names, then creating symlinks to reconstruct the repository layout. In order to get file information (package and repository metadata), we rely on simply reading a symlink. It took 3-4 months to get 4.2 TB of data, representing just 2017 (at the time of writing). Most of the information about the downloaded files and their source repository is stored in a database. In parallel, we added --- like the original `snapshot.debian.org` --- an API, [snapshot-api](https://github.com/fepitre/debian-snapshot#API), to expose information about repositories. Unlike the original one, we added much more information that rebuilder software, e.g. `debrebuild`, needs to have when requesting package information, such as the exact location of a given package in terms of Debian archive, timestamp, suite, architecture and component. The service is now publicly exposed at <https://snapshot.notset.fr> and the API endpoints at <https://snapshot.notset.fr/mr>. The service is home-hosted by the author.
```

Comment:
The 4.2 TB of data represent the whole time window from 2017 to today, not just 2017 (hopefully..)